kong/async.lua

```diff
@@ -160,5 +221,47 @@ function async:run(func, ...)
   return true
 end
 
+---
+-- Run a function repeatedly but non-overlapping
+--
+-- @tparam  number|string function execution interval (a non-zero positive number
+--                        or `"second"`, `"minute"`, `"hour"`, `"month" or `"year"`)
+-- @tparam  function      a function to run asynchronously
+-- @tparam  ...[opt]      function arguments
+-- @treturn true|nil      `true` on success, `nil` on error
+-- @treturn string|nil    `nil` on success, error message `string` on error
+function async:every(delay, func, ...)
+  delay = RECURRING[delay] or delay
+
+  assert(type(delay) == "number" and delay > 0, "invalid delay, must be number greater than zero or " ..
+                                                "'second', 'minute', 'hour', 'month' or 'year'")
+
+  local bucket = self.buckets[delay]
+  if bucket then
+    if bucket.head == BUCKET_SIZE then
+      return nil, "async bucket (" .. delay .. ") is full"
+    end
+
+  else
+    local ok, err = ngx.timer.every(delay, every_timer, self, delay)
```

Comment:
this still uses 1 timer for each "every" call right? So this would not have prevented the problem of each balancer creating its own timer and running out of them.

I would suggest to use the timerwheel, since that would run a single timer, dealing with every pattern here.

Reply:
This will create `ngx.timer.every` timer for each `delay`. One timer per delay. The functions themselves are then ran in `kong.async:run` which means they are run in parallel. Timerwheel runs functions non-parallel and implements the whole scheduling logic. Essentially to make your own `ngx.timer.every` you should perhaps add one timer that queues functions in every millisecond. That is one way around it. Still it is very uncommon to setup different delays every time with `ngx.timer.every`, and thus if you do so, I think you are using the wrong function. This library does not support `kong.async:after` (aka ngx.timer.at(>0, ...) currently.

This would fix the balancer issue as:
1. it does not allow same function to run if other run has not returned
2. if the delays are common, it will create less timers
3. it uses kong.async:run to execute the functions
4. the every timer is just a blocking loop that queues functions (it always happens a snap, so every timers are not really weight here)

There is limit in `kong.run:async`, so it will start error when you hit the limit.

I can make alternative implementation with `ngx.timer.every` that runs on each millisecond, but I feel like it adds unneccessary load to system even when that function could run very fast. The idea is not to fully eliminate the use of timers. Timers are still very valuable.

Reply:
This will create `ngx.timer.every` timer for each `delay`. One timer per delay. The functions themselves are then ran in `kong.async:run` which means they are run in parallel. Timerwheel runs functions non-parallel and implements the whole scheduling logic (which is already very nicely done with timer.every). Essentially to make your own `ngx.timer.every` you should perhaps add one timer that queues functions in every millisecond. That is one way around it. Still it is very uncommon to setup different delays every time with `ngx.timer.every`, and thus if you do so, I think you are using the wrong function. This library does not support `kong.async:after` (aka ngx.timer.at(>0, ...) currently.

This would fix the balancer issue as:
1. it does not allow same function to run if other run has not returned
2. if the delays are common, it will create less timers
3. it uses kong.async:run to execute the functions
4. the every timer is just a blocking loop that queues functions (it always happens a snap, so every timers are not really weight here)

There is limit in `kong.run:async`, so it will start error when you hit the limit.

I can make alternative implementation with `ngx.timer.every` that runs on each millisecond, but I feel like it adds unneccessary load to system even when that function could run very fast. The idea is not to fully eliminate the use of timers. Timers are still very valuable.